// textlint-rule-helper/lib/IgnoreNodeManger.js
// LICENSE : MIT
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var visit = require('unist-util-visit');
/**
 * Ignore node manager that manager ignored ranges.
 *
 */

var IgnoreNodeManger = function () {
    function IgnoreNodeManger() {
        _classCallCheck(this, IgnoreNodeManger);

        /**
         * @type {[number,number][]}
         * @private
         */
        this._ignoredRangeList = [];
    }

    /**
     * |.......|
     * ^       ^
     * Ignored Range
     *
     *    |........|
     *    ^
     *  index
     * @param {number} index
     * @returns {boolean}
     */


    _createClass(IgnoreNodeManger, [{
        key: "isIgnoredIndex",
        value: function isIgnoredIndex(index) {
            return this._ignoredRangeList.some(function (range) {
                var _range = _slicedToArray(range, 2);

                var start = _range[0];
                var end = _range[1];

                return start <= index && index <= end;
            });
        }

        /**
         * @param {[number, number]} aRange
         * @returns {boolean}
         */

    }, {
        key: "isIgnoredRange",
        value: function isIgnoredRange(aRange) {
            var index = aRange[0];
            return this.isIgnoredIndex(index);
        }

        /**
         * @param {Object} node
         * @returns {boolean}
         */

    }, {
        key: "isIgnored",
        value: function isIgnored(node) {
            var index = node.index;
            return this.isIgnoredIndex(index);
        }

        /**
         * add node to ignore range list
         * @param {TxtNode} node
         */

    }, {
        key: "ignore",
        value: function ignore(node) {
            this.ignoreRange(node.range);
        }

        /**
         * add range to ignore range list
         * @param {[number, number]} range
         */

    }, {
        key: "ignoreRange",
        value: function ignoreRange(range) {
            this._ignoredRangeList.push(range);
        }

        /**
         * ignore children node of `node`,
         * if the children node has the type that is included in `ignoredNodeTypes`.
         * @param {TxtNode} targetNode
         * @param {string[]} ignoredNodeTypes
         */

    }, {
        key: "ignoreChildrenByTypes",
        value: function ignoreChildrenByTypes(targetNode, ignoredNodeTypes) {
            var _this = this;

            visit(targetNode, function (visitedNode) {
                if (ignoredNodeTypes.indexOf(visitedNode.type) !== -1) {
                    _this.ignore(visitedNode);
                }
            });
        }
    }]);

    return IgnoreNodeManger;
}();

exports.default = IgnoreNodeManger;
