// @textlint-rule/textlint-rule-google-possessives/lib/textlint-rule-google-possessives.js
// MIT © 2017 azu
"use strict";

var _textlintReportHelperForGooglePreset = require("@textlint-rule/textlint-report-helper-for-google-preset");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DocumentURL = "https://developers.google.com/style/possessives";
var report = function report(context) {
    var Syntax = context.Syntax,
        RuleError = context.RuleError,
        getSource = context.getSource,
        fixer = context.fixer,
        report = context.report;

    var dictionaries = [
    // NG: plural word + 's
    {
        pattern: /(\w+)'s/,
        test: function test(_ref) {
            var captures = _ref.captures;

            var word = captures[0];
            // if plural word is ended in "s", ignore it.
            var isEndedS = word[word.length - 1] === "s";
            var wordPos = (0, _textlintReportHelperForGooglePreset.getPosFromSingleWord)(word);
            // Plural word
            return wordPos === "NNS" && isEndedS;
        },
        message: function message() {
            return 'A plural noun that does end in "s", add an apostrophe(\') but no additional "s" or use "of"' + "\n" + DocumentURL;
        }
    },
    // NG: singular noun + '
    {
        pattern: /([\w\s]+)'(?!s)/,
        test: function test(_ref2) {
            var captures = _ref2.captures;

            // ... word's
            // or ... the word's
            var words = captures[0].split(" ");
            var determinerWord = words[words.length - 2];
            var targetWord = words[words.length - 1];
            // if "the word's", ignore this
            if (determinerWord !== undefined) {
                var determinerType = (0, _textlintReportHelperForGooglePreset.getPosFromSingleWord)(determinerWord);
                // skip: the a
                if (determinerType === "DT") {
                    return false;
                }
            }
            var wordPos = (0, _textlintReportHelperForGooglePreset.getPosFromSingleWord)(targetWord);
            // singular noun(singular noun or Proper noun)
            return wordPos === "NNP" || wordPos === "NN";
        },
        message: function message() {
            return 'To form a possessive of a singular noun (regardless of whether it ends in s) or a plural noun that doesn\'t end in "s", add "\'s" to the end of the word' + "\n" + DocumentURL;
        }
    }];

    return _defineProperty({}, Syntax.Paragraph, function (node) {
        return (0, _textlintReportHelperForGooglePreset.paragraphReporter)({
            Syntax: Syntax,
            node: node,
            dictionaries: dictionaries,
            report: report,
            getSource: getSource,
            RuleError: RuleError,
            fixer: fixer
        });
    });
};
module.exports = {
    linter: report,
    fixer: report
};
