// @textlint/ast-node-types/lib/index.js
// MIT © 2017 azu
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * AST Node types list on TxtNode.
 * Constant value of types
 * @see https://github.com/textlint/textlint/blob/master/docs/txtnode.md
 */
var ASTNodeTypes;
(function (ASTNodeTypes) {
    ASTNodeTypes["Document"] = "Document";
    ASTNodeTypes["Paragraph"] = "Paragraph";
    ASTNodeTypes["BlockQuote"] = "BlockQuote";
    ASTNodeTypes["ListItem"] = "ListItem";
    ASTNodeTypes["List"] = "List";
    ASTNodeTypes["Header"] = "Header";
    ASTNodeTypes["CodeBlock"] = "CodeBlock";
    ASTNodeTypes["HtmlBlock"] = "HtmlBlock";
    ASTNodeTypes["ReferenceDef"] = "ReferenceDef";
    ASTNodeTypes["HorizontalRule"] = "HorizontalRule";
    ASTNodeTypes["Comment"] = "Comment";
    // inline
    ASTNodeTypes["Str"] = "Str";
    ASTNodeTypes["Break"] = "Break";
    ASTNodeTypes["Emphasis"] = "Emphasis";
    ASTNodeTypes["Strong"] = "Strong";
    ASTNodeTypes["Html"] = "Html";
    ASTNodeTypes["Link"] = "Link";
    ASTNodeTypes["Image"] = "Image";
    ASTNodeTypes["Code"] = "Code";
    ASTNodeTypes["Delete"] = "Delete";
})(ASTNodeTypes = exports.ASTNodeTypes || (exports.ASTNodeTypes = {}));
