// en-inflectors/dist/index.js
"use strict";
exports.__esModule = true;
var adjective = require("./adjective/transform");
var verb = require("./verb/transform");
var noun = require("./noun/transform");
var nounDetection = require("./noun/detect_singular_plural");
var countableDetection = require("./noun/detect_countable");
var verbsList = require("./verb/solve_lookup");
var list_uncountable_1 = require("./noun/list_uncountable");
exports.nonCountables = list_uncountable_1["default"];
var Inflectors = (function () {
    function Inflectors(word) {
        var _this = this;
        this.comparative = function () { return adjective.comparative(_this.word); };
        this.superlative = function () { return adjective.superlative(_this.word); };
        this.conjugate = function (to) { return verb.conjugate(_this.word, to); };
        this.toPresent = function () { return verb.toPresent(_this.word); };
        this.toPast = function () { return verb.toPast(_this.word); };
        this.toPastParticiple = function () { return verb.toPastParticiple(_this.word); };
        this.toPresentS = function () { return verb.toPresentS(_this.word); };
        this.toGerund = function () { return verb.toGerund(_this.word); };
        this.toPlural = function () { return noun.toPlural(_this.word); };
        this.toSingular = function () { return noun.toSingular(_this.word); };
        this.isSingular = function () { return nounDetection.isSingular(_this.word); };
        this.isPlural = function () { return nounDetection.isPlural(_this.word); };
        this.isCountable = function () { return countableDetection.isCountable(_this.word); };
        this.isNotCountable = function () { return countableDetection.isNotCountable(_this.word); };
        this.word = (word || "").toLowerCase();
    }
    return Inflectors;
}());
exports.Inflectors = Inflectors;
exports.Inflector = Inflectors;
var infinitives = verbsList.VBP;
exports.infinitives = infinitives;
