// @textlint/textlint-plugin-markdown/lib/MarkdownProcessor.js
/* eslint-disable no-unused-vars */
// LICENSE : MIT
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.MarkdownProcessor = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _markdownToAst = require("@textlint/markdown-to-ast");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var MarkdownProcessor = exports.MarkdownProcessor = function () {
    function MarkdownProcessor() {
        var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

        _classCallCheck(this, MarkdownProcessor);

        this.config = config;
        this.extensions = config.extensions ? config.extensions : [];
    }

    _createClass(MarkdownProcessor, [{
        key: "availableExtensions",
        value: function availableExtensions() {
            return [".md", ".markdown", ".mdown", ".mkdn", ".mkd", ".mdwn", ".mkdown", ".ron"].concat(this.extensions);
        }
    }, {
        key: "processor",
        value: function processor(ext) {
            return {
                preProcess: function preProcess(text, filePath) {
                    return (0, _markdownToAst.parse)(text);
                },
                postProcess: function postProcess(messages, filePath) {
                    return {
                        messages: messages,
                        filePath: filePath ? filePath : "<markdown>"
                    };
                }
            };
        }
    }]);

    return MarkdownProcessor;
}();
