// @textlint-rule/textlint-rule-google-quotation-marks/lib/textlint-rule-google-quotation-marks.js
// MIT © 2017 azu
"use strict";

var _textlintReportHelperForGooglePreset = require("@textlint-rule/textlint-report-helper-for-google-preset");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DocumentURL = "https://developers.google.com/style/quotation-marks#single-quotation-marks";
var report = function report(context) {
    var dictionaries = [
    // Commas and periods with quotation marks
    // We can not handle this rule
    // Because "Exception: When you put a keyword or other literal string in quotation marks"
    // https://developers.google.com/style/quotation-marks
    // https://developers.google.com/style/ellipses
    // {
    //     // British
    //     pattern: /"\./g,
    //     replace: () => {
    //         // American
    //         return `."`;
    //     },
    //     message: () => `Add commas and periods go inside quotation marks, in the standard American style.`
    //         + "\n"
    //         + DocumentURL
    // },
    // {
    //     pattern: /,"/g,
    //     replace: () => {
    //         return `",`;
    //     },
    //     message: () => `When you put a keyword or other literal string in quotation marks, put any other punctuation outside the quotation marks.`
    // },
    // Single quotation marks
    {
        pattern: /'([^'"]+)"([^'"]+)"([^'"]+)'/g,
        replace: function replace(_ref) {
            var captures = _ref.captures;

            return "\"" + captures[0] + "'" + captures[1] + "'" + captures[2] + "\"";
        },
        message: function message() {
            return "The outside quotation mark shoule be \", the inside quotation mark should be '.\nIn the latter case, put the primary speaker's quote in double quotation marks and the quote inside the primary speaker's quote in single quotation marks, in the standard American style. ";
        }
    }];

    var Syntax = context.Syntax,
        RuleError = context.RuleError,
        getSource = context.getSource,
        fixer = context.fixer,
        report = context.report;

    return _defineProperty({}, Syntax.Paragraph, function (node) {
        return (0, _textlintReportHelperForGooglePreset.paragraphReporter)({
            Syntax: Syntax,
            node: node,
            dictionaries: dictionaries,
            report: report,
            getSource: getSource,
            RuleError: RuleError,
            fixer: fixer
        });
    });
};
module.exports = {
    linter: report,
    fixer: report
};
